util/reggen/window.py

```diff
@@ -0,0 +1,162 @@
+                 swaccess: SWAccess):
+        assert 0 < validbits
+        assert 0 < items <= size_in_bytes
+
+        self.name = name
+        self.desc = desc
+        self.unusual = unusual
+        self.byte_write = byte_write
+        self.validbits = validbits
+        self.items = items
+        self.size_in_bytes = size_in_bytes
+        self.offset = offset
+        self.swaccess = swaccess
+
+        # Check that offset has been adjusted so that the first item in the
+        # window has all zeros in the low bits.
+        po2_size = 1 << (self.size_in_bytes - 1).bit_length()
+        assert not (offset & (po2_size - 1))
+
+    @staticmethod
+    def from_raw(offset: int,
+                 reg_width: int,
+                 params: List[Dict[str, object]],
+                 raw: object) -> 'Window':
+        rd = check_keys(raw, 'window',
+                        list(REQUIRED_FIELDS.keys()),
+                        list(OPTIONAL_FIELDS.keys()))
+
+        r_name = rd.get('name')
+        wind_desc = 'window at offset {:#x}'.format(offset)
+        if r_name is None:
+            name = None
+        else:
+            name = check_str(r_name, wind_desc)
+            wind_desc = '{!r} {}'.format(name, wind_desc)
+
+        desc = check_str(rd['desc'], 'desc field for ' + wind_desc)
+
+        unusual = check_bool(rd.get('unusual', False),
+                             'unusual field for ' + wind_desc)
+        byte_write = check_bool(rd.get('byte-write', False),
+                                'byte-write field for ' + wind_desc)
+
+        validbits = check_int(rd.get('validbits', reg_width),
+                              'validbits field for ' + wind_desc)
+        if validbits <= 0:
+            raise ValueError('validbits field for {} is not positive.'
+                             .format(wind_desc))
+        if validbits > reg_width:
+            raise ValueError('validbits field for {} is {}, '
+                             'which is greater than {}, the register width.'
+                             .format(wind_desc, validbits, reg_width))
+
+        r_items = check_str(rd['items'], 'items field for ' + wind_desc)
+        items = expand_parameter(params, r_items,
+                                 'expanding items field for ' + wind_desc)
+        if items <= 0:
+            raise ValueError("Items field for {} is {}, "
+                             "which isn't positive."
+                             .format(wind_desc, items))
+
+        assert reg_width % 8 == 0
+        size_in_bytes = items * (reg_width // 8)
+
+        # Round size_in_bytes up to the next power of 2. The calculation is
+        # like clog2 calculations in SystemVerilog, where we start with the
+        # last index, rather than the number of elements.
+        assert size_in_bytes > 0
+        po2_size = 1 << (size_in_bytes - 1).bit_length()
+
+        # A size that isn't a power of 2 is not allowed unless the unusual flag
+        # is set.
```

Comment:
Nice, thanks for adding in that flag. 

That should clear up some of the last remaining warnings that are currently blocking https://github.com/lowRISC/opentitan/pull/5026, right?

Reply:
I think there are probably a few left, but let's hope it's getting closer! :-)